app/services/allocation_copy_data_service.rb

```diff
@@ -0,0 +1,67 @@
+class AllocationCopyDataService
+  include ServicePattern
+
+  def initialize(allocation_cycle_year:)
+    @allocation_cycle_year = allocation_cycle_year
+  end
+
+  def call
+    Allocation.where(recruitment_cycle_id: previous_cycle.id).find_each do |prev_alloc|
+      next if prev_alloc.confirmed_number_of_places.to_i.zero? ||
+        allocation_exists?(prev_alloc)
+
+      provider = find_provider(prev_alloc)
+      accredited_body = find_accredited_body(prev_alloc)
+      create_allocation(prev_alloc, provider, accredited_body)
+    end
+  end
+
+private
+
+  def create_allocation(prev_alloc, provider, accredited_body)
+    Allocation.create!(
+      provider_code: prev_alloc.provider_code,
+      accredited_body_code: prev_alloc.accredited_body_code,
+      recruitment_cycle_id: current_cycle.id,
+
+      provider_id: provider.id,
+      accredited_body_id: accredited_body.id,
+
+      number_of_places: prev_alloc.confirmed_number_of_places,
+      confirmed_number_of_places: nil,
+    )
+  end
+
+  def find_provider(prev_alloc)
+    # Fetch new provider with same code for current recruitment cycle
+    Provider.where(
+      provider_code: prev_alloc.provider_code,
+      recruitment_cycle_id: current_cycle.id,
+    ).first!
+  end
+
+  def find_accredited_body(prev_alloc)
+    # Fetch new accredited_body with same code for current recruitment cycle
+    Provider.where(
+      provider_code: prev_alloc.accredited_body_code,
+      recruitment_cycle_id: current_cycle.id,
+      accrediting_provider: :accredited_body,
+    ).first!
+  end
+
+  def allocation_exists?(prev_alloc)
+    Allocation.where(
+      provider_code: prev_alloc.provider_code,
+      accredited_body_code: prev_alloc.accredited_body_code,
+      recruitment_cycle_id: current_cycle.id,
+    ).exists?
+  end
+
+  def current_cycle
+    @current_cycle ||= RecruitmentCycle.where(year: @allocation_cycle_year).first!
```

Comment:
I think there's a `RecruitmentCycle.current` method. Maybe calling this `current_cycle` is a bit ambiguous as it may not be the current it's just the cycle that matches the allocation cycle.